File: backend/app/schemas/financing.py
from __future__ import annotations

from datetime import date
from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field


class FacilityBase(BaseModel):
    property_id: UUID | None = None
    lender_type: str
    lender_name: str | None = None
    total_facility: Decimal | None = None
    opening_balance: Decimal | None = None
    rate: Decimal | None = None
    already_drawn: Decimal = Decimal("0")
    draw_eligible_override: Decimal | None = None
    requested_draw_amount: Decimal | None = None
    requested_draw_as_of: date | None = None
    commitment_source: str | None = None
    commitment_confirmed_at: datetime | None = None
    last_draw_date: date | None = None
    last_draw_amount: Decimal | None = None
    account_number: str | None = None
    account_title: str | None = None
    account_type: str | None = None
    current_balance: Decimal | None = None
    outstanding_balance: Decimal | None = None
    account_currency: str | None = None
    maturity_date: date | None = None
    member_number: str | None = None
    next_interest_payment_date: date | None = None
    next_payment_date: date | None = None
    account_nickname: str | None = None
    open_date: date | None = None
    original_loan_amount: Decimal | None = None
    payment_schedule: str | None = None
    term_length_days: int | None = None
    notes: str | None = None


class FacilityCreate(FacilityBase):
    pass


class FacilityUpdate(BaseModel):
    property_id: UUID | None = None
    lot_id: UUID | None = None
    lender_type: str | None = None
    lender_name: str | None = None
    total_facility: Decimal | None = None
    opening_balance: Decimal | None = None
    rate: Decimal | None = None
    already_drawn: Decimal | None = None
    last_draw_date: date | None = None
    last_draw_amount: Decimal | None = None
    account_number: str | None = None
    account_title: str | None = None
    account_type: str | None = None
    current_balance: Decimal | None = None
    outstanding_balance: Decimal | None = None
    account_currency: str | None = None
    maturity_date: date | None = None
    member_number: str | None = None
    next_interest_payment_date: date | None = None
    next_payment_date: date | None = None
    account_nickname: str | None = None
    open_date: date | None = None
    original_loan_amount: Decimal | None = None
    payment_schedule: str | None = None
    term_length_days: int | None = None
    notes: str | None = None


class FacilityOut(FacilityBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    lender_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class FinancingPropertyOut(BaseModel):
    property_id: UUID
    address: str
    lender_type: str
    sold_or_spec: str | None = None
    stage: str | None = None
    stage_is_estimate: bool
    possession_date: date | None = None
    build_start: date | None = None
    client_name: str | None = None
    banker_raw: str | None = None
    lender_name: str | None = None
    total_facility: Decimal | None = None
    opening_balance: Decimal | None = None
    already_drawn: Decimal | None = None
    last_draw_date: date | None = None
    last_draw_amount: Decimal | None = None
    rate: Decimal | None = None
    account_number: str | None = None
    account_title: str | None = None
    account_type: str | None = None
    current_balance: Decimal | None = None
    outstanding_balance: Decimal | None = None
    account_currency: str | None = None
    maturity_date: date | None = None
    member_number: str | None = None
    next_interest_payment_date: date | None = None
    next_payment_date: date | None = None
    account_nickname: str | None = None
    open_date: date | None = None
    original_loan_amount: Decimal | None = None
    payment_schedule: str | None = None
    term_length_days: int | None = None
    daily_interest_estimate: Decimal | None = None
    monthly_interest_estimate: Decimal | None = None
    annual_interest_estimate: Decimal | None = None
    notes: str | None = None
    draw_eligible: Decimal | None = None
    cumulative_entitled: Decimal | None = None
    funds_remaining: Decimal | None = None
    flag: str | None = None
    formula: str
    facility_id: UUID | None = None


class LenderSummary(BaseModel):
    total_drawable: Decimal | None
    properties: int
    flagged: int


class DashboardSummary(BaseModel):
    SCU: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=Decimal("0"), properties=0, flagged=0))
    PRO: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=Decimal("0"), properties=0, flagged=0))
    STRIDE: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=Decimal("0"), properties=0, flagged=0))
    RSU: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=Decimal("0"), properties=0, flagged=0))
    CLIENT: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=None, properties=0, flagged=0))
    OTHER: LenderSummary = Field(default_factory=lambda: LenderSummary(total_drawable=None, properties=0, flagged=0))


class FinancingDashboardOut(BaseModel):
    last_synced_at: datetime | None
    summary: DashboardSummary
    properties: list[FinancingPropertyOut]


class SyncResult(BaseModel):
    synced: int
    created_properties: int
    stale_deleted: int = 0
    errors: list[str]
    sync_conflicts: list[dict[str, Any]] = Field(default_factory=list)


class DocumentUploadOut(BaseModel):
    doc_id: UUID
    review_document_id: UUID | None = None
    lender_type: str
    minio_key: str
    extracted: dict[str, Any]
    requires_review: bool


class FacilityDocumentOut(BaseModel):
    id: UUID
    facility_id: UUID | None = None
    lender_type: str
    document_type: str
    minio_bucket: str
    minio_key: str
    original_filename: str | None = None
    uploaded_at: datetime
    extracted_values: dict[str, Any] | None = None
    confirmed_at: datetime | None = None
    notes: str | None = None


class ConfirmDocumentRequest(BaseModel):
    facility_id: UUID | None = None
    property_id: UUID | None = None
    values: dict[str, Any]


class ProFacilityOut(BaseModel):
    id: UUID
    facility_key: str
    property_name: str
    borrower: str | None = None
    facility_scope: str
    instrument: str | None = None
    annual_rate: Decimal | None = None
    original_advance_date: date | None = None
    original_advance_amount: Decimal | None = None
    status: str
    balance_as_of: Decimal | None = None
    last_statement_status: str | None = None
    last_statement_delta: Decimal | None = None


class ProLedgerEventOut(BaseModel):
    event_date: date
    days: int
    interest: Decimal
    draw: Decimal
    repayment: Decimal
    balance: Decimal
    accrued_interest_running_total: Decimal
    reference: str | None = None
    event_type: str


class ProLedgerOut(BaseModel):
    facility_id: UUID
    facility_key: str
    property_name: str
    as_of: date
    balance_as_of: Decimal
    events: list[ProLedgerEventOut]


class LenderStatementOut(BaseModel):
    id: UUID
    lender: str
    period: str
    minio_object_key: str
    original_filename: str | None = None
    uploaded_at: datetime
    parsed_at: datetime | None = None
    status: str


class FacilityStatementSnapshotOut(BaseModel):
    id: UUID
    statement_id: UUID
    facility_id: UUID | None = None
    matched_property_name: str
    reported_period_end_date: date
    reported_period_end_balance: Decimal
    computed_balance: Decimal | None = None
    delta: Decimal | None = None
    reconciliation_status: str
    canonical_address_key: str | None = None
    parse_payload: dict[str, Any] | None = None
    new_draws_detected: list[dict[str, Any]] | None = None


class LenderStatementDetailOut(LenderStatementOut):
    parse_payload: dict[str, Any] | None = None
    snapshots: list[FacilityStatementSnapshotOut]


class ClientDrawScheduleItem(BaseModel):
    seq: int
    label_raw: str
    stage_key: str | None = None
    amount: Decimal | None = None
    amount_type: str = "fixed"
    percent: Decimal | None = None
    conditions_raw: str | None = None
    source_page: int | None = None


class ClientDrawDeposit(BaseModel):
    seq: int | None = None
    label_raw: str | None = None
    amount: Decimal | None = None
    due_raw: str | None = None
    source_page: int | None = None


class ClientDrawScheduleOut(BaseModel):
    id: UUID
    property_id: UUID
    document_id: UUID
    minio_object_key: str
    original_filename: str | None = None
    purchase_price: Decimal | None = None
    client_name: str | None = None
    otp_date: date | None = None
    schedule: list[dict[str, Any]]
    deposits: list[dict[str, Any]]
    extraction_confidence: str
    extraction_status: str
    extraction_notes: str | None = None
    reviewed_by: UUID | None = None
    reviewed_at: datetime | None = None
    superseded_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ClientDrawScheduleReviewRequest(BaseModel):
    purchase_price: Decimal | None = None
    client_name: str | None = None
    otp_date: date | None = None
    schedule: list[dict[str, Any]]
    deposits: list[dict[str, Any]] = Field(default_factory=list)
    extraction_notes: str | None = None


class ClientDrawRequestOut(BaseModel):
    id: UUID
    property_id: UUID
    schedule_id: UUID
    draw_items: list[dict[str, Any]]
    amount: Decimal
    stage_at_prep: str | None = None
    prepared_at: datetime
    prepared_by: UUID | None = None
    status: str
    notes: str | None = None


class ClientPrepDrawConfirmRequest(BaseModel):
    draw_items: list[dict[str, Any]]
    amount: Decimal
    notes: str | None = None


class ClientDrawStatusRequest(BaseModel):
    status: str
    notes: str | None = None


class ClientPrepDrawOut(BaseModel):
    status: str
    property: dict[str, Any]
    current_stage: str | None = None
    current_stage_synced_at: datetime | None = None
    schedule: ClientDrawScheduleOut | None = None
    schedule_table: list[dict[str, Any]] = Field(default_factory=list)
    requestable_items: list[dict[str, Any]] = Field(default_factory=list)
    already_requested_items: list[dict[str, Any]] = Field(default_factory=list)
    unmapped_items: list[dict[str, Any]] = Field(default_factory=list)
    next_upcoming_item: dict[str, Any] | None = None
    requestable_total: Decimal | None = None
    eligibility_unavailable_reason: str | None = None
    lawyer_note: str | None = None

File: frontend/src/types/financing.ts
export type LenderType = "SCU" | "PRO" | "STRIDE" | "RSU" | "CLIENT" | "OTHER";
export type DrawFlag = "OVER_DRAWN" | "FACILITY_NOT_SET" | "NOT_STARTED" | "CHECK_OTP" | "NO_PROGRESS_REPORT" | "NEEDS_LINK" | "SYNC_CONFLICT" | null;

export interface LenderSummary {
  total_drawable: string | number | null;
  properties: number;
  flagged: number;
}

export interface FinancingProperty {
  property_id: string;
  address: string;
  lender_type: LenderType;
  sold_or_spec?: string | null;
  stage?: string | null;
  stage_is_estimate: boolean;
  possession_date?: string | null;
  build_start?: string | null;
  client_name?: string | null;
  banker_raw?: string | null;
  lender_name?: string | null;
  total_facility?: string | number | null;
  opening_balance?: string | number | null;
  already_drawn?: string | number | null;
  last_draw_date?: string | null;
  last_draw_amount?: string | number | null;
  requested_draw_amount?: string | number | null;
  requested_draw_as_of?: string | null;
  commitment_source?: string | null;
  commitment_confirmed_at?: string | null;
  rate?: string | number | null;
  account_number?: string | null;
  account_title?: string | null;
  account_type?: string | null;
  current_balance?: string | number | null;
  outstanding_balance?: string | number | null;
  account_currency?: string | null;
  maturity_date?: string | null;
  member_number?: string | null;
  next_interest_payment_date?: string | null;
  next_payment_date?: string | null;
  account_nickname?: string | null;
  open_date?: string | null;
  original_loan_amount?: string | number | null;
  payment_schedule?: string | null;
  term_length_days?: number | null;
  daily_interest_estimate?: string | number | null;
  monthly_interest_estimate?: string | number | null;
  annual_interest_estimate?: string | number | null;
  notes?: string | null;
  draw_eligible?: string | number | null;
  cumulative_entitled?: string | number | null;
  funds_remaining?: string | number | null;
  flag: DrawFlag;
  formula: string;
  facility_id?: string | null;
}

export interface ClientDrawSchedule {
  id: string;
  property_id: string;
  document_id: string;
  minio_object_key: string;
  original_filename?: string | null;
  purchase_price?: string | number | null;
  client_name?: string | null;
  otp_date?: string | null;
  schedule: Array<Record<string, unknown>>;
  deposits: Array<Record<string, unknown>>;
  extraction_confidence: "high" | "needs_review" | string;
  extraction_status: string;
  extraction_notes?: string | null;
  reviewed_at?: string | null;
  superseded_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClientDrawRequest {
  id: string;
  property_id: string;
  schedule_id: string;
  draw_items: Array<Record<string, unknown>>;
  amount: string | number;
  stage_at_prep?: string | null;
  prepared_at: string;
  status: "prepared" | "sent_to_lawyer" | "funded" | "cancelled" | string;
  notes?: string | null;
}

export interface ClientPrepDrawPackage {
  status: "needs_otp" | "stage_unavailable" | "ready" | string;
  property: Record<string, unknown>;
  current_stage?: string | null;
  current_stage_synced_at?: string | null;
  schedule?: ClientDrawSchedule | null;
  schedule_table: Array<Record<string, unknown>>;
  requestable_items: Array<Record<string, unknown>>;
  already_requested_items: Array<Record<string, unknown>>;
  unmapped_items: Array<Record<string, unknown>>;
  next_upcoming_item?: Record<string, unknown> | null;
  requestable_total?: string | number | null;
  eligibility_unavailable_reason?: string | null;
  lawyer_note?: string | null;
}

export interface FinancingDashboard {
  last_synced_at?: string | null;
  summary: Record<LenderType, LenderSummary>;
  properties: FinancingProperty[];
}

export interface FacilityPayload {
  property_id: string;
  lender_type: LenderType;
  lender_name?: string | null;
  total_facility?: string | number | null;
  opening_balance?: string | number | null;
  rate?: string | number | null;
  already_drawn?: string | number | null;
  draw_eligible_override?: string | number | null;
  requested_draw_amount?: string | number | null;
  requested_draw_as_of?: string | null;
  commitment_source?: string | null;
  commitment_confirmed_at?: string | null;
  last_draw_date?: string | null;
  last_draw_amount?: string | number | null;
  account_number?: string | null;
  account_title?: string | null;
  account_type?: string | null;
  current_balance?: string | number | null;
  outstanding_balance?: string | number | null;
  account_currency?: string | null;
  maturity_date?: string | null;
  member_number?: string | null;
  next_interest_payment_date?: string | null;
  next_payment_date?: string | null;
  account_nickname?: string | null;
  open_date?: string | null;
  original_loan_amount?: string | number | null;
  payment_schedule?: string | null;
  term_length_days?: number | string | null;
  notes?: string | null;
}

export interface FacilityAssignmentPayload {
  facility_type: LenderType;
  lender_id: string;
  total_facility?: string | number | null;
  opening_balance?: string | number | null;
  rate?: string | number | null;
  already_drawn?: string | number | null;
  draw_eligible_override?: string | number | null;
  requested_draw_amount?: string | number | null;
  requested_draw_as_of?: string | null;
  commitment_source?: string | null;
  commitment_confirmed_at?: string | null;
  notes?: string | null;
}

export interface FacilityRecord extends FacilityPayload {
  id: string;
  lender_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface UploadResponse {
  doc_id: string;
  review_document_id?: string | null;
  lender_type: LenderType;
  minio_key: string;
  extracted: Record<string, unknown>;
  requires_review: boolean;
}

export interface ProFacility {
  id: string;
  facility_key: string;
  property_name: string;
  borrower?: string | null;
  facility_scope: "lot" | "development" | string;
  instrument?: string | null;
  annual_rate?: string | number | null;
  original_advance_date?: string | null;
  original_advance_amount?: string | number | null;
  status: string;
  balance_as_of?: string | number | null;
  last_statement_status?: string | null;
  last_statement_delta?: string | number | null;
}

export interface ProLedgerEvent {
  event_date: string;
  days: number;
  interest: string | number;
  draw: string | number;
  repayment: string | number;
  balance: string | number;
  accrued_interest_running_total: string | number;
  reference?: string | null;
  event_type: string;
}

export interface ProLedger {
  facility_id: string;
  facility_key: string;
  property_name: string;
  as_of: string;
  balance_as_of: string | number;
  events: ProLedgerEvent[];
}

export interface LenderStatement {
  id: string;
  lender: string;
  period: string;
  minio_object_key: string;
  original_filename?: string | null;
  uploaded_at: string;
  parsed_at?: string | null;
  status: string;
}

export interface FacilityStatementSnapshot {
  id: string;
  statement_id: string;
  facility_id?: string | null;
  matched_property_name: string;
  reported_period_end_date: string;
  reported_period_end_balance: string | number;
  computed_balance?: string | number | null;
  delta?: string | number | null;
  reconciliation_status: string;
  canonical_address_key?: string | null;
  parse_payload?: Record<string, unknown> | null;
  new_draws_detected?: Array<Record<string, unknown>> | null;
}

export interface LenderStatementDetail extends LenderStatement {
  parse_payload?: Record<string, unknown> | null;
  snapshots: FacilityStatementSnapshot[];
}
